Microsoft.Toolkit.Mvvm/ComponentModel/ViewModelBase.cs

```diff
@@ -0,0 +1,182 @@
+        /// </remarks>
+        protected ViewModelBase()
+            : this(Messaging.Messenger.Default)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ViewModelBase"/> class.
+        /// </summary>
+        /// <param name="messenger">The <see cref="IMessenger"/> instance to use to send messages.</param>
+        protected ViewModelBase(IMessenger messenger)
+        {
+            Messenger = messenger;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="IMessenger"/> instance in use.
+        /// </summary>
+        protected IMessenger Messenger { get; }
+
+        private bool isActive;
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the current view model is currently active.
+        /// </summary>
+        public bool IsActive
+        {
+            get => this.isActive;
+            set
+            {
+                if (Set(ref this.isActive, value, true))
+                {
+                    if (value)
+                    {
+                        OnActivated();
+                    }
+                    else
+                    {
+                        OnDeactivated();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Raised whenever the <see cref="IsActive"/> property is set to <see langword="true"/>.
+        /// Use this method to register to messages and do other initialization for this instance.
+        /// </summary>
+        protected virtual void OnActivated()
+        {
+        }
+
+        /// <summary>
+        /// Raised whenever the <see cref="IsActive"/> property is set to <see langword="false"/>.
+        /// Use this method to unregister from messages and do general cleanup for this instance.
+        /// </summary>
+        /// <remarks>The base implementation unregisters all messages for this recipient (with no token).</remarks>
+        protected virtual void OnDeactivated()
+        {
+            Messenger.Unregister(this);
+        }
+
+        /// <summary>
+        /// Broadcasts a <see cref="PropertyChangedMessage{T}"/> with the specified
+        /// parameters, without using any particular token (so using the default channel).
+        /// </summary>
+        /// <typeparam name="T">The type of the property that changed.</typeparam>
+        /// <param name="oldValue">The value of the property before it changed.</param>
+        /// <param name="newValue">The value of the property after it changed.</param>
+        /// <param name="propertyName">The name of the property that changed.</param>
+        /// <remarks>
+        /// You should override this method if you wish to customize the channel being
+        /// used to send the message (eg. if you need to use a specific token for the channel).
+        /// </remarks>
+        protected virtual void Broadcast<T>(T oldValue, T newValue, string propertyName)
```

Comment:
I'll make an example from my app Brainf*ck#.
I have what I call a "compact memory viewer" in my shell, which is a grid-style visualizer of the current machine state, updated in real time as you execute commands in the console. [Here's a screen](https://user-images.githubusercontent.com/10199417/87182891-535b1980-c2e5-11ea-9a48-b8177b00ee67.png)

In my `ConsoleViewModel` I have the following property:

```csharp
private IReadOnlyMachineState _MachineState;

/// <summary>
/// Gets the <see cref="IReadOnlyMachineState"/> instance currently in use
/// </summary>
public IReadOnlyMachineState MachineState
{
    get => _MachineState;
    private set => Set(ref _MachineState, value, true);
}
```

You can see I'm using the `Set` method with the `bool broadcast` parameter (from the `ViewModelBase` class), which in turn will call that `Broadcast` method to notify other modules in the app about the property change. I then have a `CompactMemoryViewerViewModel` class which listens to `PropertyChangedMessage<IReadOnlyMachineState>` messages, and upon receiving one will update the collection bound to the UI, which is displayed in that control.

This is particularly handy to make the code less verbose, as you won't have to declare lots of different single-purpose messages, but you'll be able to just listen to your desired `PropertyChangedMessage<T>` type.

Another possible example - listen to a `PropertyChangedMessage<User>` message to broadcast when the currently logged user changes in an application that supports multiple users that can be switched to at any time.

As for why the method is `virtual` - that's so that users are able to customize that behavior if they wanted to, eg. by sending the broadcast message with a specific token instead, or to also do some additional logging before sending each broadcast message, etc. It really depends on the situation, the point here is to make the base type as flexible as possible.

Reply:
I get it, cool. We should have a sample showing using the new `ISubscriber` paradigm for this too, think that'd be really cool. Thanks!